src/data/java/com/github/minecraftschurli/arsmagicalegacy/data/AMAdvancementProvider.java

```diff
@@ -0,0 +1,122 @@
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.level.ItemLike;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Set;
+import java.util.function.Consumer;
+
+@SuppressWarnings("unchecked")
+class AMAdvancementProvider extends AdvancementProvider {
+    private static final Logger LOGGER = LogManager.getLogger();
+    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+    private final DataGenerator generator;
+
+    AMAdvancementProvider(DataGenerator pGenerator) {
+        super(pGenerator);
+        generator = pGenerator;
+    }
+
+    @Override
+    public void run(HashCache pCache) {
+        Path path = generator.getOutputFolder();
+        Set<ResourceLocation> set = Sets.newHashSet();
+        Consumer<Advancement> consumer = a -> {
+            if (!set.add(a.getId())) throw new IllegalStateException("Duplicate advancement " + a.getId());
+            else {
+                Path path1 = path.resolve("data/" + a.getId().getNamespace() + "/advancements/" + a.getId().getPath() + ".json");
+                try {
+                    DataProvider.save(GSON, pCache, a.deconstruct().serializeToJson(), path1);
+                } catch (IOException e) {
+                    LOGGER.error("Couldn't save advancement {}", path1, e);
+                }
+            }
+        };
+        for (Consumer<Consumer<Advancement>> c : ImmutableList.of(new AMAdvancements(), new AMBookAdvancements())) c.accept(consumer);
+    }
+
+    /**
+     * Adds an advancement.
+     *
+     * @param consumer The advancement consumer provided by the data generator.
+     * @param folder   The folder name of this advancement's tab.
+     * @param id       The id of this advancement.
+     * @param parent   The id of this advancement's parent.
+     * @param item     The item to display.
+     * @param type     The frame type of this advancement.
+     * @param toast    Whether to show a toast in the top right corner or not.
+     * @param chat     Whether to announce the advancement to chat or not.
+     * @param hide     Whether the advancement should be visible before unlocking or not.
+     * @param criteria Optional advancement criteria. Should be made through Pair.of("id", criterion);
+     */
+    static Advancement addAdvancement(Consumer<Advancement> consumer, String folder, String id, Advancement parent, ItemLike item, FrameType type, boolean toast, boolean chat, boolean hide, Pair<String, CriterionTriggerInstance>... criteria) {
+        Advancement.Builder builder = Advancement.Builder.advancement().parent(parent).display(item, new TranslatableComponent("advancements." + ArsMagicaAPI.MOD_ID + "." + id + ".title"), new TranslatableComponent("advancements." + ArsMagicaAPI.MOD_ID + "." + id + ".description"), null, type, toast, chat, hide);
```

Comment:
use `Util.makeDescriptionId()` instead
```suggestion
        Advancement.Builder builder = Advancement.Builder.advancement().parent(parent).display(item, new TranslatableComponent(Util.makeDescriptionId("advancement", new ResourceLocation(ArsMagicaAPI.MOD_ID, id))+".title"), new TranslatableComponent(Util.makeDescriptionId("advancement", new ResourceLocation(ArsMagicaAPI.MOD_ID, id))+".description"), null, type, toast, chat, hide);
```